drivers/bm/bm_dfu_client.c

```diff
@@ -0,0 +1,439 @@
+    bm_dfu_event_ack_received_t ack_evt;
+    uint8_t tx_buf[sizeof(bm_frame_header_t) + sizeof(bm_dfu_event_ack_received_t) + sizeof(bm_dfu_frame_header_t)];
+
+    /* Stuff BM Frame Header*/
+    frm_hdr.version = BM_V0;
+    frm_hdr.payload_type = BM_DFU;
+    frm_hdr.payload_length = sizeof(bm_dfu_event_ack_received_t) + sizeof(bm_dfu_frame_header_t);
+
+    /* Stuff ACK Event */
+    ack_evt.success = success;
+    ack_evt.err_code = err_code;
+
+    memcpy(tx_buf, &frm_hdr, sizeof(bm_frame_header_t));
+	tx_buf[sizeof(bm_frame_header_t)] = BM_DFU_ACK;
+    memcpy(&tx_buf[sizeof(bm_frame_header_t) + sizeof(bm_dfu_frame_header_t)], (uint8_t *) &ack_evt, sizeof(ack_evt));
+    
+    ack_frm = (bm_frame_t *)tx_buf;
+    if (bm_serial_frm_put(ack_frm))
+    {
+        LOG_ERR("ACK not sent");
+    }
+}
+
+static void bm_dfu_client_abort(void)
+{
+    bm_frame_header_t frm_hdr;
+    bm_frame_t *abort_frm;
+    uint8_t tx_buf[sizeof(bm_frame_header_t) + sizeof(bm_dfu_frame_header_t)];
+
+    /* Stuff BM Frame Header*/
+    frm_hdr.version = BM_V0;
+    frm_hdr.payload_type = BM_DFU;
+    frm_hdr.payload_length = sizeof(bm_dfu_frame_header_t);
+
+    memcpy(tx_buf, &frm_hdr, sizeof(bm_frame_header_t));
+	tx_buf[sizeof(bm_frame_header_t)] = BM_DFU_ABORT;
+
+    abort_frm = (bm_frame_t *)tx_buf;
+    if (bm_serial_frm_put(abort_frm))
+    {
+        LOG_ERR("ABORT not sent");
+    }
+}
+
+static void bm_dfu_client_update_end(uint8_t success, uint8_t err_code)
+{
+    bm_frame_header_t frm_hdr;
+    bm_frame_t *update_end_frm;
+    bm_dfu_event_update_end_t update_end_evt;
+
+    uint8_t tx_buf[sizeof(bm_frame_header_t) + sizeof(bm_dfu_event_update_end_t) + sizeof(bm_dfu_frame_header_t)];
+
+    /* Stuff BM Frame Header*/
+    frm_hdr.version = BM_V0;
+    frm_hdr.payload_type = BM_DFU;
+    frm_hdr.payload_length = sizeof(bm_dfu_event_update_end_t) + sizeof(bm_dfu_frame_header_t);
+
+    /* Stuff Update End Event */
+    update_end_evt.success = success;
+    update_end_evt.err_code = err_code;
+
+    memcpy(tx_buf, &frm_hdr, sizeof(bm_frame_header_t));
+	tx_buf[sizeof(bm_frame_header_t)] = BM_DFU_END;
+    memcpy(&tx_buf[sizeof(bm_frame_header_t) + sizeof(bm_dfu_frame_header_t)], (uint8_t *) &update_end_evt, sizeof(update_end_evt));
+
+    update_end_frm = (bm_frame_t *)tx_buf;
+    if (bm_serial_frm_put(update_end_frm))
+    {
+        LOG_ERR("DFU End not sent");
+        return;
+    }
+}
+
+static void chunk_timer_handler(struct k_timer *tmr)
+{
+    bm_dfu_event_t evt;
+
+    evt.type = DFU_EVENT_CHUNK_TIMEOUT;
+    if(k_msgq_put(_client_context.dfu_subystem_queue, &evt, K_NO_WAIT))
+    {
+        LOG_ERR("Message could not be added to Queue");
+    }
+}
+
+static int bm_dfu_process_payload(uint16_t man_decode_len, uint8_t * man_decode_buf)
+{
+    int retval = 0;
+
+    /* Account for BM Frame Header size and BM_DFU payload type*/
+    uint16_t _dfu_frame_len = man_decode_len - sizeof(bm_frame_header_t) - sizeof(bm_dfu_frame_header_t);
```

Comment:
Should we be checking that `man_decode_len > sizeof(bm_frame_header_t) - sizeof(bm_dfu_frame_header_t)`?